[MCA] New option to report scheduling information: -scheduling-info

jvillette38 opened the pull request:
This is a new way to update scheduling information in llvm. I have used this to update scheduling information for AArch64 Neoverse V1 micro architecture (new patches will follow and will be dependent to this pull request).

This pull request contains 2 commits:
A) `llvm-mca -scheduling-info` option
B) `update_mca_test_checks.py` new options: `--check-sched-info` and `--update-sched-info`.

A) `llvm-mca -scheduling-info` disables default llvm-mca reporting (InstructionInfoView) and output information in the following format:
`
<uOps> | <Latency>  | <Bypass Latency> | <Throughput> | <Resources> | <LLVM Opcode>  | <Assembly input: instruction + comment>
`
Example from new llvm-mca test `AArch64/Neoverse/V1-scheduling-info.s`:
Input:
`
  abs  v25.2s, v25.2s  // ABS <Vd>.<T>, <Vn>.<T>  \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV
`
Output:
`
1 | 2 | 2 | 4.00 | V1UnitSVE01, V1UnitV | ABSv2i32 | abs  v25.2s, v25.2s  // ABS <Vd>.<T>, <Vn>.<T>  \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV
`

So if we are able to extract scheduling information from micro architecture document for each instruction variant, it is possible to write test in this form and check `llvm-mca -scheduling-info` output for the differences between llvm information compared to the one in comments. If you get differences, check the documentation to update comment or fix llvm to update llvm-mca output.
LLVM Opcode is given to make easier the changes in target description.

B) `update_mca_test_checks.py  --check-sched-info` is used to check informations between `llvm-mca` output and information in comments. If found differences, it will exit with error code and report them. Developer can fix comments or llvm target description or use `update_mca_test_checks.py --update-sched-info` to update automatically comments and then check differences with git.

Convention for comments used by new update_mca_test_checks.py options:
- C or C++ style comment: '/* */' and '//'
- Fields:
`
<asm instruction> <// or /*> <instruction format> \\ <micro architecture reference> \\ <uOps> <Latency> <Bypass latency> <Throughput> <Resources seperated with commas>
`

@mshockwave and @Rin18 may be interested.

JulienVillette committed:
[mca] update_mca_test_checks.py helpers to manage llvm-mca -scheduling-info new option

When using llvm-mca -scheduling-info and if assembly
test contains comments with reference values of scheduling
info: <MicroOps> <Latency> <Forward Latency> <Throughput> <Units>

To check coherency between llvm-mca -scheduling-info output
and scheduling references in comment, use --check-sched-info.
Exit with error if found deferences and report them.

This is usefull to check new scheduling info patches as
we can specify source documentation references for each
instructions and so be able to understand easier differences.

Example of comment in AArch64/Neoverse/V1-scheduling-info.s:
  abs D15, D23  // ABS <V><d>, <V><n> \\ ASIMD arith, basic \\ 1 2 2 4.0 V1UnitV

llvm-mca -scheduling-info output:
 1 2 2 4.00 - ABSv1i64 V1UnitSVE01, V1UnitV,
  abs d15, d23  // ABS <V><d>, <V><n> \\ ASIMD arith, basic \\ 1 2 2 4.0 V1UnitV

update_mca_test_checks.py is searching for 4 values at the begining and
after comment // and compare these values. Values order must be the same
as llvm-mca output. And it will check that all resources in comment
(reference) is included in llvm-mca output.

It is possible to update source test scheduling information references
using -update-sched-info option. If you want to update test source
references and llvm-mca output references, you have to run two times
update_mca_test_checks.py -update-sched-info. First time to update
scheduling information references and second time to update llvm-mca new
output reference.

JulienVillette committed:
[mca] New option -scheduling-info

Outputs micro ops, latency, bypass latency, throughput, llvm opcode
name, used resources and parsed assembly instruction with comments.

This option is used to compare scheduling info from micro architecture documents.
Reference scheduling information (from Architecture and micro
architecture) are in comment section after each instruction (// or
/* */).

These information may be generated from Architecture Description Language.
By this way, it is easy to compare information from llvm and from
documentation/ADL.

LLVM Opcode name help to find right instruction regexp to fix in
Target Scheduling Info specification.

Example:
  Input:      abs D20, D11  // ABS <V><d>, <V><n>
      \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV
  Output:  1 | 2 | 2 | 4.00 | V1UnitV | ABSv1i64 |
           abs     d20, d11              // ABS <V><d>, <V><n>
      \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV

c-rhodes commented on llvm/docs/CommandGuide/llvm-mca.rst:
```diff
@@ -170,6 +170,20 @@ option specifies "``-``", then the output will also be sent to standard output.
+  comments to check easily reported and reference scheduling information.
+  Suggested information in comment:
+  // <architecture instruction form> \\ <scheduling documentation title> \\
+     <uOps>, <Latency>, <Bypass Latency>, <Throughput>, <Resources units>
```
still failing to build, see llvm/docs/README.txt for instructions on how to build locally. Please could you also run the formatters?

jvillette38 replied:
It should be ok now regarding coding style checking.

c-rhodes reviewed (commented):
thanks for the patch. IIUC, this adds a new `-scheduling-info` option, which is similar to `-instruction-info`. So whereas `-instruction-info` has:
```
Instruction Info:
[1]: #uOps
[2]: Latency
[3]: RThroughput
[4]: MayLoad
[5]: MayStore
[6]: HasSideEffects (U)
```

this option has:
```
Scheduling Info:
[1]: #uOps
[2]: Latency
[3]: Bypass Latency
[4]: Throughput
[5]: Resources
[6]: LLVM OpcodeName
[7]: Instruction
[8]: Comment if any
```

and the purpose of this new option is to make it easier to compare the model against a good reference (and fix if necessary), which is specified via a comment after the assembly instruction.

So for example:

```
echo "abs D15, D11  /* ABS <V><d>, <V><n>  \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV */" | build/bin/llvm-mca -mtriple=aarch64 -mcpu=neoverse-v1 -scheduling-info
...
Scheduling Info:
[1]: #uOps
[2]: Latency
[3]: Bypass Latency
[4]: Throughput
[5]: Resources
[6]: LLVM OpcodeName
[7]: Instruction
[8]: Comment if any
 [1]    [2]  [3]   [4]      [5]                                                                [6]                [7]                                 [8]
  1    | 2  | 2   | 4.00   | V1UnitV                                                          | ABSv1i64         | abs  d15, d11                       /* ABS <V><d>, <V><n>  \ ASIMD arith, basic  \ 1 2  2  4.0 V1UnitV */
```

here we can compare `1    | 2  | 2   | 4.00   | V1UnitV` against the reference in the comment and see it's ok.

There's a few things going on here. I can see value in some of the extra info alongside the instruction like bypass latency/resources/LLVM OpcodeName, but this could be added to the existing `-instruction-info` view if there's agreement this is useful?

For the reference, why not just make the CHECK lines in a test like `llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-basic-instructions.s` be the reference? It would have to be XFAIL'ed until the model matched the reference, but any difference would be obvious looking at the diff after running the update script.

boomanaiden154 reviewed (commented):
Haven't looked in detail at the code yet, but I'm not necessarily sure `llvm-mca` is the best place for this. For validating/updating scheduling models, we already have `llvm-exegesis`, where this would be a more natural fit.

As a side note, I don't often have much confidence in manufacturer-provided scheduling information. It could be that ARM's information is much better than what we get from AMD/Intel, but on X86/RISCV so far, using `llvm-exegesis` with benchmarks to validate scheduling information has been more effective than trying to construct scheduling models from manufacturer reference documentation.

jvillette38 commented:
> There's a few things going on here. I can see value in some of the extra info alongside the instruction like bypass latency/resources/LLVM OpcodeName, but this could be added to the existing `-instruction-info` view if there's agreement this is useful?
> 
> For the reference, why not just make the CHECK lines in a test like `llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-basic-instructions.s` be the reference? It would have to be XFAIL'ed until the model matched the reference, but any difference would be obvious looking at the diff after running the update script.

Thanks for the review.
I can add this information in InstructionInfoView:
```
[3]: Bypass Latency
[4]: Throughput
[5]: Resources
[6]: LLVM OpcodeName
[7]: Instruction
[8]: Comment if any
```

I can let reverse throughput instead of throughput (it was easier to compare with documentation). The goal was to have an easy way to compare llvm scheduling information with documentation/llvm-exegesis and be able to fix llvm target description with  LLVM opcode name.

Note: If I do these changes in InstructionInfoView, it will modify all tests that use this view.
Are you agree with that?

My point was to extract scheduling information from documentation or llvm-exegesis to get a reference and then fix easily llvm scheduling information in target description (I have done that by using Software Optimization Guide). l should use llvm-exegesis in second time.
So It is possible to use comments with scheduling information from documentation/llvm-exegesis during update of target description (scripts help to check differences) and then we can remove them when pushing updates.

jvillette38 commented:
> Haven't looked in detail at the code yet, but I'm not necessarily sure `llvm-mca` is the best place for this. For validating/updating scheduling models, we already have `llvm-exegesis`, where this would be a more natural fit.
> 
> As a side note, I don't often have much confidence in manufacturer-provided scheduling information. It could be that ARM's information is much better than what we get from AMD/Intel, but on X86/RISCV so far, using `llvm-exegesis` with benchmarks to validate scheduling information has been more effective than trying to construct scheduling models from manufacturer reference documentation.

I agree with you but it is a first start and llvm-exegesis is here to refine these information. This patch is here to help to quickly update llvm target description from scheduling information reference (documentation and/or llvm-exegesis).

michaelmaitland commented on llvm/tools/llvm-mca/Views/SchedulingInfoView.cpp:
```diff
@@ -0,0 +1,212 @@
+    const MCWriteProcResEntry *Last = STI.getWriteProcResEnd(&SCDesc);
+    auto sep = "";
+    for (; Index != Last; ++Index) {
+      if (!Index->ReleaseAtCycle)
```
Does AcquireAtCycles need to be handled?

jvillette38 replied:
No because the goal is to get reservation cycles to compute the throughput.
https://github.com/llvm/llvm-project/blob/main/llvm/lib/MC/MCSchedule.cpp#L97
But for now, AcquireAtCycle does not appear and cannot be checked regarding scheduling modele...

michaelmaitland commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-scheduling-info.s:
It's also unclear to me how this test case was constructed. How could we generate this kind of test for other scheduler models?

boomanaiden154 replied:
`llvm-exegesis` can do this with `-opcode-index=-1` (assuming exegesis supports generating benchmarks for those opcodes). It wouldn't be difficult to implement something like that in `llvm-mca`.

https://github.com/llvm/llvm-project/blob/89d636ba91924c945d39f02f261863b709249744/llvm/tools/llvm-exegesis/llvm-exegesis.cpp#L335-L343

jvillette38 replied:
Thanks for the review.
I have generated this test by parsing ARM architecture + Software Optimization Guide...
Next step is to get more precise information with llvm-exegesis.
In all cases, I think it is a good starting point to have complete list of instructions (architecture document) and then get scheduling information from micro-architecture document or even better from llvm-exegesis...

boomanaiden154 commented:
> I agree with you but it is a first start and llvm-exegesis is here to refine these information. This patch is here to help to quickly update llvm target description from scheduling information reference (documentation and/or llvm-exegesis).

Originally my point was that this functionality should be part of exegesis rather than being in MCA, but I can see how this information would also potentially be useful to end users, so I guess I'm fine with it either way.

JulienVillette committed:
[MCA] SchedulingInfoView: consider negative ReadAdvanceCycle and AcquireAtCycle

MR #126703:
 - Negative ReadAdvance cycles can be negative, so add ForwardingDelayCycles to Latency (computeInstrLatency).

RKSimon commented:
I must agree with @boomanaiden154 on this - I'd expect a better approach to be to generate llvm-exegesis yaml for each instruction and use the analysis tools to compare against the scheduler models. 

I do something similar in my "uops_to_exegesis" script here: https://github.com/RKSimon/llvm-scripts

jvillette38 commented on llvm/lib/MC/MCSchedule.cpp:
```diff
@@ -174,3 +174,40 @@ MCSchedModel::getForwardingDelayCycles(ArrayRef<MCReadAdvanceEntry> Entries,
+    return 0;
+
+  unsigned Latency = 0;
+  unsigned maxLatency = 0;
```
Done

jvillette38 commented on llvm/lib/MC/MCSchedule.cpp:
```diff
@@ -174,3 +174,40 @@ MCSchedModel::getForwardingDelayCycles(ArrayRef<MCReadAdvanceEntry> Entries,
+    const MCWriteLatencyEntry *WLEntry =
+        STI.getWriteLatencyEntry(&SCDesc, DefIdx);
+    // Early exit if we found an invalid latency.
+    // Consider no bypass
```
Fixed and only filter bad latencies by clamping latency value to 0.

mshockwave reviewed (commented):
> I must agree with @boomanaiden154 on this - I'd expect a better approach to be to generate llvm-exegesis yaml for each instruction and use the analysis tools to compare against the scheduler models.

Majority of the RISCV MCA test are actually there to make sure our TableGen changes on scheduling models reflect correctly, especially on vector instructions which use lots of non-trivial TableGen machinery. 

So while I agree that a llvm-exegesis-generated YAML should be a better golden reference for scheduling models, I think the existing MCA tests still have their values.

In general, I think printing just `<uOps>, <Latency>, <Bypass Latency>, <Throughput>, <Resources units>` for scheduling model tests makes sense to me, as the MCA view we use in scheduling model tests right now is a little too verbose.
But I would expect `update_mca_tests_checks.py` to generate something like
```
// CHECK: <uOps>, <Latency>, <Bypass Latency>, <Throughput>, <Resources units>
```
Which effectively does the same things as the current scheduling model tests, that is, doing the validation as part of the test, rather than doing the validation using the script via `--check-sched-info` as you propose.

jvillette38 commented:
New MR #128892 to integrate instructions from V1-scheduling-info.s test in existing AArch64 Neoverse V1. Just added one test: V1-misc-instructions.s for undocumented instructions in ARM Software Optimization Guide.
If MR #128892 is accepted, I will create new MR for llvm-mca -scheduling-info option implemented in InstructionInfoView.
And then I will close actual MR (#126703).

jvillette38 commented:
Integrated with MRs #128892 #130574 #132972.
